Replace some uses of Poco::Path with std::filesystem::path

jclarkeSTFC opened the pull request:
I've replaced some usages of `Poco::Path` with `std::filesystem::path`, and in some places where strings were being used to represent paths I switched to `std::filesystem::path`. I made changes to `ConfigService` so would be good to give this a thorough check. See #37868 for further details on this migration.

Note: `Poco::Path::expand` does not have an equivalent in the standard library, it expands all environment variables in the path.

One gotcha I found, if you thought that `Poco::Path::makeDirectory` would make a directory, then you'd be wrong, it removes the filename from the `Poco::Path` object.

`Poco::Path` objects know if they represent a file or a directory (they have some internal flag), but there's no way to tell if a `std::filesystem::path` is supposed to point at a file or a directory. The only way I could think of was to check if the path has an extension or not, but `Poco::TemporaryFile` returns a file with no extension, and that's used in a number of tests.

<!-- Why has this work been done? If there is no linked issue please provide appropriate context for this work.
#### Purpose of work
This can be removed if a github issue is referenced below
-->

Refs #37868. <!-- and fix #xxxx or close #xxxx xor resolves #xxxx. One line per issue fixed. -->

### To test:

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

*This does not require release notes* because **this is an internal refactor, API should stay the same.**

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

jclarkeSTFC committed:
Replace some uses of Poco::Path with std::filesystem::path

Poco::Path::expand does not have an equivalent in the standard
library.

jclarkeSTFC committed:
Fix cppcheck warnings

jclarkeSTFC committed:
Return std::strings instead of std::filesystem::paths

To maintain the same Python API, return std::strings when getting
directories from ConfigService.

jclarkeSTFC committed:
Poco::Path::makeDirectory does not make a directory

Turns out it just removes any filename from the path object, so
the equivalent is std::filesystem::path::remove_filename.

jclarkeSTFC committed:
Refactor out several paths represented as strings

Representing a path as a string causes pain because you have to
worry about whether or not there is a slash, which direction is
the slash on which OS, etc.

jclarkeSTFC committed:
Replace use of Poco::Path to fix slash dependence

Poco::Path::resolve did not like the std::filesystem::path not having
a slash at the end (!), so I removed Poco to make this method use
std instead.

jclarkeSTFC committed:
Change logic to match Poco::Path::isDirectory

Poco::Path::isDirectory only checks if the Poco::Path object has something
labelled as a filename, not necessarily if it exists. Hence to get the
same functionality we need to check if the path exists or not, and also
if it is a directory or not. If it's an existing file then it's no
good.

jclarkeSTFC committed:
Fixes for Windows compilation

Msbuild apparently is not happy with creating Poco::File objects
from std::filesystem::path so I gave it some help.

jclarkeSTFC committed:
Fixes following change from Poco::Path to std::filesystem::path

We shouldn't use a string to represent a file path, because then we
have to worry about slashes and which way they're going, etc.

jclarkeSTFC committed:
Windows needs help converting a path to a string

Msbuild requires that .string() on std::filesystem::path be called in a
variety of places, whereas Clang and GCC don't have a problem going
from a std::filesystem::path to a Poco::Path via a string.

sf1919 commented:
As this is failing tests and further commit(s) are expected we won't re-run any tests on the new Linux nodes.

jclarkeSTFC committed:
Fix bad conversions from Poco::Path or string

In LoadEventNexus the last component of the instrument directory was
being lost in the conversion from Poco::Path to std::filesystem::path.
The easiest way to fix this was to just use std::filesystem::path
instead.

Similar conversion problem in GroupingLoader, same fix.

jclarkeSTFC committed:
Use path instead of Poco::Path in IDFObject

It was simpler to do this than put in special code for handling the
change between Poco::Path and std::filesystem::path.

jclarkeSTFC committed:
This should be creating an absolute path

jclarkeSTFC committed:
Use generic strings in LoadRawHelper

This aids comparisons across different operating systems, e.g. on
Windows you'll get a backslash for a path separator.

jclarkeSTFC committed:
Use Path comparison instead of strings

More robust on different operating systems, don't have to worry about
path separators.

jclarkeSTFC committed:
Use path comparison in tests

We don't care whether we have forward slashes or backslashes in the
path, so we use a path comparison instead of a string comparison.

jclarkeSTFC committed:
Remove some uses of filesystem::path::absolute

The previous implementation using Poco::Path resolve() is not the same
as absolute(). There's is also a fix for isFile() checking for whether
the Poco::Path object has a file name, not whether the actual file
exists.

jclarkeSTFC committed:
Keep old behaviour from Poco::Path

Creating a Poco::Path with an empty string then calling .parent() will
give you '../', so we need to keep this behaviour with filesystem::path
for backwards compatibility.

jclarkeSTFC committed:
Use has_filename() instead of extension

Temporary files do not have an extension, especially those generated
by Poco::TemporaryFile, which is used a lot in tests.
